### PR TITLE
feature: add Cursor#setReadConcern builder method (3.6)

### PR DESCRIPTION
**Pull request by reggi**
NODE-2806

### CODE REVIEW

File: lib/cursor.js [hunk 1]
```diff
@@ -768,6 +769,29 @@ class Cursor extends CoreCursor {
...
   }
 
+  /**
+   * Set the ReadPreference for the cursor.
+   * @method
+   * @param {(string|ReadConcern)} readConcern The new read preference for the cursor.
+   * @throws {MongoError}
+   * @return {Cursor}
+   */
+  setReadConcern(readConcern) {
```
Comment (mbroadst) [inline lib/cursor.js:779]: nit: I think that `withReadConcern` reads better, thoughts?

File: lib/cursor.js [hunk 2]
```diff
@@ -768,6 +769,29 @@ class Cursor extends CoreCursor {
...
+      throw new MongoError('cannot change cursor readConcern after cursor has been accessed');
+    }
+
+    if (readConcern instanceof ReadConcern) {
+      this.options.readConcern = readConcern;
+    } else if (typeof readConcern === 'string') {
+      this.options.readConcern = new ReadConcern(readConcern);
+    } else {
+      throw new TypeError('Invalid read concern: ' + readConcern);
+    }
```
Comment (mbroadst) [inline lib/cursor.js:790]: I wonder if this logic belongs on the `ReadConcern` class itself, perhaps `ReadConcern.resolve(readConcernLike)`? It looks like `ReadConcern.fromOptions` does a similar check

File: test/functional/cursor.test.js [hunk 3]
```diff
@@ -2839,6 +2840,28 @@ describe('Cursor', function() {
     }
   });
 
+  /**
+   * @ignore
+   */
+  it('shouldSetReadConcern', {
```
Comment (mbroadst) [inline test/functional/cursor.test.js:2846]: nit: this compactedCamelCaseForm is from the legacy `integra` runner. Let's call this something more readable for future maintainers. 
```suggestion
  it('should allow setting the readConcern through a builder method', {
```
or something like that

File: test/functional/cursor.test.js [hunk 4]
```diff
@@ -2839,6 +2840,28 @@ describe('Cursor', function() {
...
+    metadata: {
+      requires: { topology: ['single', 'replicaset', 'sharded', 'ssl', 'heap', 'wiredtiger'] }
+    },
+    test: function(done) {
```
Comment (mbroadst) [inline test/functional/cursor.test.js:2850]: should we use `withClient` here?

File: test/functional/cursor.test.js [hunk 5]
```diff
@@ -2839,6 +2840,28 @@ describe('Cursor', function() {
     }
   });
 
+  /**
+   * @ignore
+   */
+  it('shouldSetReadConcern', {
+    metadata: {
+      requires: { topology: ['single', 'replicaset', 'sharded', 'ssl', 'heap', 'wiredtiger'] }
```
Comment (mbroadst) [inline test/functional/cursor.test.js:2848]: if you want to test on all variants then there is no need to provide this `metadata` at all.

**Review by mbroadst — CHANGES_REQUESTED** [review]

File: lib/cursor.js (hunk 1 above)
Comment (reggi) [reply lib/cursor.js:779]: I think it should be consistent with `setReadPreference`.

File: lib/cursor.js (hunk 2 above)
Comment (reggi) [reply lib/cursor.js:790]: We've already handled this in master and have pretty good `fromOptions` methods there, I can port those over, I just thought we'd be anti the back-porting. [In the master branch I re-wrote this to use `.fromOptions` for both `setReadPreference` and `setReadConcern`.](https://github.com/mongodb/node-mongodb-native/pull/2584/files#diff-f7a848a5828483a0958ec4e436ca7f277d846a669c610b5647e192450331d09dR754-R773)

Let me know if you're interested in a back-port of those classes.

File: test/functional/cursor.test.js (hunk 3 above)
Comment (reggi) [reply test/functional/cursor.test.js:2846]: Cool, good to know.

File: test/functional/cursor.test.js (hunk 4 above)
Comment (reggi) [reply test/functional/cursor.test.js:2850]: Yeah I'm down.

File: test/functional/cursor.test.js (hunk 5 above)
Comment (reggi) [reply test/functional/cursor.test.js:2848]: cool also good to know.

File: lib/cursor.js (hunk 2 above)
Comment (reggi) [reply lib/cursor.js]: done, backported updated .fromOptions from both classes and have these methods same as master.

File: lib/cursor.js (hunk 1 above)
Comment (mbroadst) [reply lib/cursor.js:779]: I'm thinking we should change `setReadPreference` to `withReadPreference` :smile:

File: lib/cursor.js (hunk 2 above)
Comment (mbroadst) [reply lib/cursor.js]: Ok, I think it's a good start to backport the `fromOptions` changes, but I'm still wondering if we can improve things a bit. It seems like generally in both `ReadConcern.fromOptions` and this new method that we want to be able to take something that is `ReadConcernLike`, convert it to a `ReadConcern` instance, and throw an error if the provided value doesn't satisfy the requirements of `ReadConcernLike`. It seems to me like an intermediate method could be used in both places (I suggested `resolve` above), and if we had that method we could avoid the extra allocation here when passing an object to `fromOptions` in this method.

File: lib/cursor.js (hunk 1 above)
Comment (reggi) [reply lib/cursor.js:779]: This is 3.6 that would be a breaking change here, are you ok with that?

File: test/functional/cursor.test.js [hunk 6]
```diff
@@ -9,6 +9,8 @@ const Buffer = require('safe-buffer').Buffer;
 const Writable = require('stream').Writable;
 
 const core = require('../../lib/core');
+const ReadConcern = require('../../lib/read_concern');
+const { withClient } = require('./shared');
```
Comment (emadum) [inline test/functional/cursor.test.js:13]:
```suggestion
const withClient = require('./shared').withClient;
```

File: lib/read_concern.js [hunk 7]
```diff
@@ -29,11 +29,14 @@ class ReadConcern {
     }
 
     if (options.readConcern) {
-      if (options.readConcern instanceof ReadConcern) {
-        return options.readConcern;
+      const { readConcern } = options;
```
Comment (emadum) [inline lib/read_concern.js:32]:
```suggestion
      const readConcern = options.readConcern;
```

**Review by emadum — APPROVED** [review]
Some destructuring which needs to be taken out since this is `3.6`, other than that LGTM 👍

File: lib/cursor.js (hunk 1 above)
Comment (mbroadst) [reply lib/cursor.js:779]: I was thinking in 3.6 we would provide both, and deprecate `setReadPreference`, advising users to use `withReadPreference` instead

File: test/functional/shared.js [hunk 8]
```diff
@@ -148,7 +148,7 @@ function withClient(client, callback) {
     }
     return client
       .connect()
-      .then(callback)
+      .then(callback.bind(this))
```
Comment (mbroadst) [inline test/functional/shared.js:151]: do we need to make this change in `master` as well?

File: lib/core/topologies/read_preference.js [hunk 9]
```diff
@@ -100,20 +100,23 @@ const VALID_MODES = [
...
   if (typeof readPreference === 'string') {
     return new ReadPreference(readPreference, readPreferenceTags);
   } else if (!(readPreference instanceof ReadPreference) && typeof readPreference === 'object') {
     const mode = readPreference.mode || readPreference.preference;
     if (mode && typeof mode === 'string') {
       return new ReadPreference(mode, readPreference.tags, {
-        maxStalenessSeconds: readPreference.maxStalenessSeconds || maxStalenessSeconds,
-        hedge: readPreference.hedge
+        maxStalenessSeconds: readPreference.maxStalenessSeconds,
+        hedge: options.hedge
```
Comment (mbroadst) [inline lib/core/topologies/read_preference.js:119]: what motivated these changes? It looks like we've lost support for `maxStalenessSeconds` from `options`, and I believe `hedge` is not a valid connection string option

**Review by mbroadst — CHANGES_REQUESTED** [review]

File: lib/cursor.js (hunk 2 above)
Comment (reggi) [reply lib/cursor.js]: I've gone down a rabbit hole to refactor and fix `ReadConcern` and `ReadPreference`. ReadPreference seems pretty buggy and there's a lot of duplication. And I also would love to replace `ReadPreference.translate` and use a `options = resolve()` in-place. This request here has a bit of scope-creep which is why I first implemented it the way I did. Complications arise when deciding default values, validation, returning undefined / throwing / default, and parental inheritance, and some type guard functions. I kind of have an opinionated way of how all this should be done. Also I think there's a bit of overlap with Neal's client options work as well.  I'm not sure what to do here, I don't want to respond with an unexpected massive refactor of these classes. Part of me just wants to rip the band-aid off and fix them.

File: test/functional/shared.js (hunk 8 above)
Comment (reggi) [reply test/functional/shared.js:151]: yes

File: lib/core/topologies/read_preference.js (hunk 9 above)
Comment (reggi) [reply lib/core/topologies/read_preference.js:119]: yes your right, this is in draft now.

Comment (reggi) [reply lib/core/topologies/read_preference.js:119]: I pulled changes from master thinking that had the updated code, seems like master is wrong and needs updates reflected in 3.6

Comment (reggi) [reply lib/core/topologies/read_preference.js:119]: Will pull in these changes in master

File: test/functional/cursor.test.js (hunk 6 above)
Comment (reggi) [reply test/functional/cursor.test.js:13]: 👍 new linter should help with this

**Review by nbbeeken — APPROVED** [review]

**Pull request closed without merging**